design/core_collection_operations.md

```diff
@@ -0,0 +1,175 @@
+    version
+  * Similar to the collections version, it also allows for easy rollbacks, and any version
+    semantics are up to the user.
+  * In the future, we could add a schema version if the schema needs to change. For now, no
+    schema version implies a schema version of 1.
+    * An altenative would just be to add a version to the product name, like `taxa_freq2`.
+
+```
+[
+    {
+        "product": "taxa_freq",
+        "version": "r207.kbase.3"
+    },
+    {
+        "product": "genome_stats",
+        "version': "r207.kbase.14"
+    }
+]
+```
+
+## Implementation
+
+### DB schema
+
+* The collections document will be essentially as represented above, with the keys / values
+  as described and a list of associated data products.
+* There will be two separate collections - `coll_all_versions` (or something) and
+  `coll_active_versions`.
+  * The former holds all versions of all collections, and the latter holds only the active
+    version of each collection.
+* Activating a version means:
+  * Pulling the collection version document from `coll_all_versions`
+  * Setting the activation date to the current date
+  * Overwriting the collection document in `coll_active_versions`.
+    * Atomic!
+* Alternatives:
+  * Tag the active version in `coll_active_version`.
+    * There's a fundamental problem with this approach in that it's non-atomic - at some point
+      either no versions will be tagged as active or > 1 will.
+    * This can be alleviated with transactions but in Arango [transactions are only reliable in
+      non-sharded environments](https://github.com/arangodb/arangodb/issues/11424).
+    * Furthermore, filtering data becomes much slower since all the inactive versions need to
+      be filtered out. Index size goes up because you need (at least) the activity state in
+      the equality part of the index. If you want to filter inactive versions, you need a
+      separate index without the activity state in the equality part of the index.
+    * Also generally more complex than the proposed schema.
+  * Have the service manage the collection version with a monotonically increasing version, like
+    `workspace_deluxe` or `sample_service`
+    * This approach is much more complex and is not necessary for the use case proposed here. It
+      also has the filtering drawback as mentioned above.
+
+### Endpoints
+
+* Only includes immediately relevant endpoints.
+
+#### List collections
+
+* For now returns all active collections based on the < 1000 collections assumption.
+
+```
+GET /collections
+```
+
+#### Save a collection
+
+* Requires a token with the `COLLECTIONS_ADMIN` (or something) `auth2` role.
+
+```
+POST /collections/{collection_id}/versions/{version}
+
+BODY contains collection contents.
+```
+
+#### Activate a collection
+
+* Requires a token with the `COLLECTIONS_ADMIN` (or something) `auth2` role.
+
+```
+PUT /collections/{collection_id}/versions/{version}/activate
+```
```

Comment:
Sorry, more nits I just spotted. `PUT` doesn't feel appropriate here as running `GET` on the same URL will not return the "PUTted" resource. More of a GET or a POST (if a body is required) IMO

Reply:
I'm using PUT because PUTs are idempotent and POSTS are not. GET isn't appropriate because the endpoint can result in a state change.

Reply:
`PUT` is pretty explicitly supposed to replace the contents at a resource. `GET` feels most appropriate here, excluding the elusive `PATCH`, which is probably correct but tended not to be well supported as-of the last time I worked on a REST API.

Reply:
> GET feels most appropriate here

The problem with a GET here, as I mentioned, is that this causes a state change, which GET requests are not supposed to do

Reply:
update, the more I think about `GET` the less correct it feels

Reply:
Could there be a case for "deactivating" the active version (so as to hide a collection) with `PUT /collections/{collection_id}/versions/{version}/activate`?

Also, potentially in the future: `GET /collections/{collection_id}/versions/{version}/activate` returns a bool
seems fine-ish? 

happiest path seems to be:
```HTTP
PUT /collections/{collection_id}/versions/{version}/activate
Content-Type: application/json

<boolean>
```
```
200 OK
```
Later:
```HTTP
GET /collections/{collection_id}/versions/{version}/activate
```
``` 
200 OK

<boolean>
```

Reply:
By process of elimination:
HEAD, OPTIONS - yeah, no
GET - this op does a state change, so no
PATCH - means we're replacing part of an entity, which is not what's happening here
POST - means we're creating a new entity, which is not what's happening here
PUT - means we're replacing an entity, which is not what's happening here exactly, but we *are* replacing the *active* collection entity

To me PUT looks like the least bad choice

Reply:
> Could there be a case for "deactivating" the active version (so as to hide a collection) with PUT

Hmm, my KJR is that once a collection is published it shouldn't be unpublished... if this is an operation we want I think this probably needs discussion with the PO.

And wouldn't it be a DELETE?

Reply:
> Also, potentially in the future: GET /collections/{collection_id}/versions/{version}/activate returns a bool
seems fine-ish?

I'm not sure what the purpose of that would be, since the active collection (and version) can be retrieved via
GET /collections/{collection_id}

Reply:
I'm not sure what the purpose of returning a bool from the PUT is - it'll always be `true` unless there's an error, in which case the user will get an entirely different error structure

Reply:
Resolution per out of band conversations:
* We'll go with PUT
* We won't worry about the deactivation stuff (including PUT bodies, GET etc.) for now and cross that bridge when we come to it